File: 207.course_search/canFinish.js
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {boolean}
 */

var canFinish = function(numCourses, prerequisites) {
    // create graph
    var graph = [];
    for(let i = 0; i < numCourses; i++){
        graph[i] = [];
    }
    var visited = [];
    for(let i = 0; i < numCourses; i++){
        visited[i] = 0;
    }
    //
    prerequisites.forEach(function(x){
        graph[x[0]].push(x[1]);
    });

    var dfs = function(graph, visited, i){
        if(visited[i] === -1){//if ith node is marked as being visited, then a cycle is found
            return false;
        }
        if(visited[i] === 1){//if it is done visted, then do not visit again
            return true;
        }
        visited[i] = -1;

        for(let j = 0; j < graph[i].length; j++){
            if(!dfs(graph,visited,graph[i][j])){
                return false;
            }
        }
        visited[i] = 1;
        return true;
    };

    for(let i = 0; i < numCourses; i++){
        if(!dfs(graph, visited, i)){
            return false;
        }
    }
    return true;
};

File: 207.course_search/canFinish.py
from collections import defaultdict
class Solution(object):
    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """
        def dfs(graph, visited, i):
            if visited[i] == -1: # there is cycle detected
                return False
            if visited[i] == 1: # has been visited
                return True 
            visited[i] = -1 # marked as being visited
            for j in graph[i]:
                if not dfs(graph, visited, j):
                    return False
            visited[i] = 1 # marked as visited
            return True
        visited = [0] * numCourses
        graph = defaultdict(list)
        # initialize the graph
        for pair in prerequisites:
            graph[pair[0]].append(pair[1])
        for i in xrange(numCourses):
            if not dfs(graph, visited, i):
                return False
        return True

'''
use dfs to check if there is cycle inside the graph
'''